Update to `pnpm@10` and `node@22`

pauliesnug opened the pull request:
i also standardized formatting (added a few new rules like import/export sorting and some cool style tricks, but can revert them) (or revert the apply formatting commit and run it for easier diffing) everything seems to work/run well. :3

fixes #223

NotNite reviewed (changes requested):
I do want to update pnpm and Node.js in the project, and pnpm catalogs are a very welcome addition, but there shouldn't be any styling changes in this PR. Please drop `60b1517955df00f78cc37aa34b6914ebc059e863` and `983b86f28141d9f9db9d83733cea8a4373478208`.

Also, please don't rewrite the ESLint config or build scripts (at least right now). ESLint config changes should go into https://github.com/moonlight-mod/eslint-config since it's shared across multiple repositories (but please make an issue/discussion before PRing any changes there!). Prettier configs are ??? and I'm not entirely sure what to do about them yet. I don't think the build script benefits from being TypeScript right now, but maybe we could in the future.

(also hi again thank you for contributing)

pauliesnug commented:
> I do want to update pnpm and Node.js in the project, and pnpm catalogs are a very welcome addition, but there shouldn't be any styling changes in this PR. Please drop `60b1517955df00f78cc37aa34b6914ebc059e863` and `983b86f28141d9f9db9d83733cea8a4373478208`.
> 
> Also, please don't rewrite the ESLint config or build scripts (at least right now). ESLint config changes should go into https://github.com/moonlight-mod/eslint-config since it's shared across multiple repositories (but please make an issue/discussion before PRing any changes there!). Prettier configs are ??? and I'm not entirely sure what to do about them yet. I don't think the build script benefits from being TypeScript right now, but maybe we could in the future.
> 
> (also hi again thank you for contributing)

haii okay! ill make an issue for the styling commit and chery-pick it later. does the typescript buildscript change make a diff? since its same logic? i can still scrap it tho!

NotNite commented:
I'd prefer to remove the TypeScript build script only because it doesn't benefit from strict typing a lot. There's not a lot of objects that need typing and it doesn't get changes very often, so it doesn't make a lot of sense to require compiling it (even if you're using tsx or that one experimental node flag). I also wanted to unify it with our *other* esbuild config (there's too many configs lol) in the future, so having it in JS right now is easier.

(also yes you can message me there)

pauliesnug commented:
> I'd prefer to remove the TypeScript build script only because it doesn't benefit from strict typing a lot. There's not a lot of objects that need typing and it doesn't get changes very often, so it doesn't make a lot of sense to require compiling it (even if you're using tsx or that one experimental node flag). I also wanted to unify it with our _other_ esbuild config (there's too many configs lol) in the future, so having it in JS right now is easier.
> 
> (also yes you can message me there)

Does it work if I unify it in this monorepo or should it be a separate repository? Also reverted the formatting.

NotNite commented:
Thanks for reverting that! Just a few other things to touch up on:

- Make sure to add back the Prettier configs!
- I would preferably have the scripts still be JavaScript (for similar reasons to the build script), please revert that. Also, please remove `tsx` and `taze` as dependencies, we won't need `tsx` anymore and I don't want to add `taze` in this PR (but it does look neat).
- I'm not sure why, but `packages/core/src/extension.ts` was changed for some reason. Could you revert that?

> Does it work if I unify it in this monorepo or should it be a separate repository? Also reverted the formatting. not sure why but the reverted formatting file takes forever to lint

Ideally it would still be in a separate repository, I don't like having shared things in this repository because wrangling build systems & CI is annoying (with the exception of `@moonlight-mod/types` since that's just part of the codebase anyways).

pauliesnug commented:
> Thanks for reverting that! Just a few other things to touch up on:
> 
>     * Make sure to add back the Prettier configs!
> 
>     * I would preferably have the scripts still be JavaScript (for similar reasons to the build script), please revert that. Also, please remove `tsx` and `taze` as dependencies, we won't need `tsx` anymore and I don't want to add `taze` in this PR (but it does look neat).
> 
>     * I'm not sure why, but `packages/core/src/extension.ts` was changed for some reason. Could you revert that?
> 
> 
> > Does it work if I unify it in this monorepo or should it be a separate repository? Also reverted the formatting. not sure why but the reverted formatting file takes forever to lint
> 
> Ideally it would still be in a separate repository, I don't like having shared things in this repository because wrangling build systems & CI is annoying (with the exception of `@moonlight-mod/types` since that's just part of the codebase anyways).
cool, i added taze as a replacement for `update.mjs` because `pnpm update` doesn't work with catalogs yet

pauliesnug commented:
> Thanks for reverting that! Just a few other things to touch up on:
> 
>     * Make sure to add back the Prettier configs!
> 
>     * I would preferably have the scripts still be JavaScript (for similar reasons to the build script), please revert that. Also, please remove `tsx` and `taze` as dependencies, we won't need `tsx` anymore and I don't want to add `taze` in this PR (but it does look neat).
> 
>     * I'm not sure why, but `packages/core/src/extension.ts` was changed for some reason. Could you revert that?
> 
> 
> > Does it work if I unify it in this monorepo or should it be a separate repository? Also reverted the formatting. not sure why but the reverted formatting file takes forever to lint
> 
> Ideally it would still be in a separate repository, I don't like having shared things in this repository because wrangling build systems & CI is annoying (with the exception of `@moonlight-mod/types` since that's just part of the codebase anyways).
cool, i added taze as a replacement for `update.mjs` because `pnpm update` doesn't work with catalogs yet

NotNite commented:
> cool, i added taze as a replacement for update.mjs because pnpm update doesn't work with catalogs yet

Makes sense. I'm fine with keeping it then, but you can still remove it if you want.

NotNite commented:
Almost there, just a few more changes that I just realized last second (sorry):

- `esbuild-copy-static-files` got uninstalled, it was a dev dependency for the build script. Build script doesn't build without it, so please add it back. (That plugin is extremely bad anyways but that's a topic for another day.)
- This PR is targeted to `main` and I am stupid and didn't notice it until now. You can re-target it to develop by hitting the edit button, then resolve merge conflicts if there are any (probably just me bumping a package).

Other than that, I think this is ready for merge. Get those two minor things resolved and I'll go test it one last time. Thank you so much!

NotNite commented:
Thanks! I'll go test this a bit and fix up some nitpicks, then I'll let CI run and see if it works.

NotNite commented:
Seems like CI works fine, there's just one small problem I've noticed with moving everything to catalogs. When referencing the types package directly (`"@moonlight-mod/types": "github:pauliesnug/moonlight#pnpmv10&path:packages/types"`), it fails to install since it uses the catalog protocol:

```
 ERR_PNPM_SPEC_NOT_SUPPORTED_BY_ANY_RESOLVER  @moonlight-mod/lunast@catalog:prod isn't supported by any available resolver.

This error happened while installing the dependencies of @moonlight-mod/types@1.3.14

An external package outside of the pnpm workspace declared a dependency using
the catalog protocol. This is likely a bug in that external package. Only
packages within the pnpm workspace may use catalogs. Usages of the catalog
protocol are replaced with real specifiers on 'pnpm publish'.

This is likely a bug in the publishing automation of this package. Consider filing
a bug with the authors of:

  @moonlight-mod/types@1.3.14

Progress: resolved 271, reused 249, downloaded 1, added 0
```

I think the types package will need to have the versions duplicated :(. We need to support importing from `file:` or `github:` URIs for people doing local development or using types that aren't published yet. We miss out on synchronizing the dependencies for mappings, but meh. This same restriction also probably applies to `workspace:`, we just didn't use it there.

Let's remove the unique dependencies from the catalog (e.g. `standalone-electron-types` but not `@moonlight-mod/mappings`), copy the versions into the types package.json, and see if it works. Also we should probably remove `engines`/`engineStrict` from there, this package is used in every moonlight extension and I don't wanna make extension developers upgrade immediately.

Feel free to do this if you're willing, else I'll do it in the morning (make sure to pull my changes though). You can test it by creating an extension (`pnpm create @moonlight-mod/extension`) and then referencing your local moonlight repository (`file:../moonlight/packages/types`).

NotNite reviewed (approved):
LGTM, merging as a squash commit and hoping nothing breaks. :rocket:

I wanted to move a lot sooner but I just didn't have the energy to fix everything that could possibly break. I've tested everything I can think of (referencing types directly, dry-run of pnpm publish, messing with builds, ESLint, etc.) and it seems stable. I appreciate your patience and speed with my several requested changes. I'll also have to handle some docs changes to reflect the update, and maybe I'll add the update script back(?), but that doesn't have to be in this PR.

In the future, if you want to touch the esbuild/ESLint configs, please let me know beforehand (issue/discussion, post in the moonlight Discord server, DM me, whatever). I don't want to entirely rewrite the config or heavily change formatting/linting rules, and I don't want to close a PR that had a lot of time spent on it. In general, ask beforehand before making any major changes (and please scope down your PRs in the future lol).

(Also, on a more personal note, it's fun to work on PRs with my friends. Thanks for contributing here, I appreciate it :D)

pauliesnug commented:
swag :3 >< !!
